core/channel-upgrade.go

```diff
@@ -303,6 +304,77 @@ func NewUpgradeState(chanState chantypes.State, upgradeExists bool) (UpgradeStat
 	}
 }
 
+type queryUpgradeChannelStateResult struct {
+	updateHeaders        []Header
+	channel              *chantypes.QueryChannelResponse
+	chanUpg              *chantypes.QueryUpgradeResponse
+	settled              bool
+	upgradeState         UpgradeState
+	proposedConnectionID string
+	upgErr               *chantypes.QueryUpgradeErrorResponse
+}
+
+func queryUpgradeChannelState(ctx context.Context, sh SyncHeaders, prover, counterparty *ProvableChain) (*queryUpgradeChannelStateResult, error) {
+	var ret queryUpgradeChannelStateResult
+	logger := GetChannelPairLoggerRelative(prover, counterparty)
+	queryCtx := sh.GetQueryContext(ctx, prover.ChainID())
+
+	var err error
+	ret.updateHeaders, err = sh.SetupHeadersForUpdate(ctx, prover, counterparty)
+	if err != nil {
+		logger.ErrorContext(ctx, "failed to set up headers for LC update on both chains", err)
+		return nil, err
+	}
+
+	ret.channel, ret.settled, err = querySettledChannel(queryCtx, logger, prover, true)
+	if err != nil {
+		logger.ErrorContext(ctx, "failed to query the channel with proofs", err)
+		return nil, err
+	} else if !ret.settled {
+		return &ret, nil
+	}
+
+	ret.chanUpg, ret.settled, err = querySettledChannelUpgrade(queryCtx, logger, prover, true)
+	if err != nil {
+		logger.ErrorContext(ctx, "failed to query the channel upgrade pair with proofs", err)
+		return nil, err
+	} else if !ret.settled {
+		return &ret, nil
+	}
+
+	ret.upgradeState, err = NewUpgradeState(ret.channel.Channel.State, ret.chanUpg != nil)
+	if err != nil {
+		logger.ErrorContext(ctx, "failed to create UpgradeState", err)
+		return nil, err
+	}
+
+	/**
+	 * proposedConnectionID is used only if
+	 * - counterparty's action is UPGRADE_ACTION_TRY, it is only if  my upgradeState is UPGRADE_STATE_INIT
+	 * - call upgradeAlreadyComplete, it is only if my upgradeState is UPGRADE_STATE_FLUSHCOMPLETE
+	 */
+	if ret.upgradeState == UPGRADE_STATE_INIT || ret.upgradeState == UPGRADE_STATE_FLUSHCOMPLETE {
+		ret.proposedConnectionID, err = queryProposedConnectionID(queryCtx, prover, ret.chanUpg)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	/**
+	 * upgErr is used only if
+	 * - counterparty's actions is UPGRADE_ACTION_CANCEL or UPGRADE_ACTION_CANCEL_FLUSHCOMPLETE
+	 * - my upgradeState is UPGRADE_STATE_UNINIT or UPGRADE_STATE_INIT
+	 */
+	if ret.upgradeState == UPGRADE_STATE_UNINIT || ret.upgradeState == UPGRADE_STATE_INIT {
+		ret.upgErr, err = QueryChannelUpgradeError(queryCtx, prover, true)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
If `QueryChannelUpgradeError` (for Ethereum) can't find an error event, it traverses events backwards ... up to the genesis block!
Therefore we must not call this function under overly broad conditions.

Fortunately, `QueryChannelUpgradeError` only accesses blockchain events, not blockchain states, and so we need not call this in `queryUpgradeChannelState`.